### PR TITLE
ForeignPngFilter

### PR DESCRIPTION
**Pull request by randyridge**
add ForeignPngFilter

### CODE REVIEW

**Review by kleisauke — CHANGES_REQUESTED** [review]
Thanks! I've added a few inline suggestions, but otherwise this looks good to me! The reason that this "enum" was not [semi-generated](https://github.com/kleisauke/net-vips/blob/ef078c9fb3dfc3b2ef05741ea8da842c18b28b9b/samples/NetVips.Samples/Samples/GenerateEnums.cs) is because `VipsForeignPngFilter` is not an enum, but a bit field.

@jcupitt Do you think the other bindings can also benefit from this? I had a similar issue with wasm-vips, see for example:
https://github.com/kleisauke/pyvips/tree/flags-helper
https://github.com/kleisauke/wasm-vips/blob/master/build/gen_type_declarations.py#L228-L230
https://github.com/kleisauke/wasm-vips/blob/master/build/gen_type_declarations.py#L251-L253
https://github.com/kleisauke/wasm-vips/blob/master/build/gen_type_declarations.py#L283

File: src/NetVips/Enums.cs [hunk 1]
```diff
@@ -450,6 +450,30 @@ public static class ForeignHeifCompression
             public const string Av1 = "av1";
         }
 
+        /// <summary>
+        /// The PNG filter to use.
+        /// </summary>
+        public static class ForeignPngFilter
```
[the remark this answers is not in the document]
Comment (kleisauke) [reply src/NetVips/Enums.cs:457]: Oh whoops, that flag can of course not be applied to a static class.

**Comment by jcupitt** [comment]
Yes, the flags/enum thing is used in some places, but not worked through all the bindings. It might be useful to fix this.

**Review by kleisauke — APPROVED** [review]

**Comment by kleisauke** [comment]
Thank you Randy, this will be in v1.2.5.

@jcupitt As far as I know `VipsForeignPngFilter` is the only flag/bitfield used in libvips. I noticed that it is also missing in pyvips.

**Comment by jcupitt** [comment]
There are a few unimportant ones, eg.:

```
john@yingna ~/GIT/libvips/libvips/include/vips (master) $ grep "/*< flag" *.h
foreign.h:typedef enum /*< flags >*/ {
foreign.h:typedef enum /*< flags >*/ {
object.h:typedef enum /*< flags >*/ {
operation.h:typedef enum /*< flags >*/ {
```

gtk-doc uses the magic `/*< .. >*/` comment to spot them, and you can find them by introspection.

I wonder if `VipsForeignPngFilter` is a true bitfield. Does libpng really support `avg | paeth`? I suspect it behaves more like a simple enum, though I'm too lazy to check.

**Comment by kleisauke** [comment]
I'll also add the corresponding `Enums.ForeignPngFilter? filter = null` overload for `Pngsave{,Buffer,Target,Stream}` allowing you to do this (for example):
```csharp
var im = Image.NewFromFile("lichtenstein.jpg", access: Enums.Access.Sequential);
im.Pngsave("lichtenstein.png", filter: /*(int)*/Enums.ForeignPngFilter.All);
```
(it currently requires casting an `enum` to `int`, which is a bit clumsy)

**Comment by kleisauke** [comment]
@jcupitt I'm not not entirely sure if you could do `avg | paeth`, but `none | sub | up | avg | paeth` is possible (which is just a synonym for `all`). Anyways, it's defined here:
https://github.com/glennrp/libpng/blob/dbe3e0c43e549a1602286144d94b0666549b18e6/png.h#L1459-L1471

**Comment by kleisauke** [comment]
I can confirm that `avg | paeth` also works. It generates a PNG file with exclusively `avg` or `paeth` as row filter, see for example:
```bash
$ pngcheck -vv lichtenstein.png | grep filters -A 1 | tail -5
    row filters (0 none, 1 sub, 2 up, 3 avg, 4 paeth):
      3 4 4 4 4 4 4 (2415 out of 2421)
--
    row filters (0 none, 1 sub, 2 up, 3 avg, 4 paeth):
      4 3 4 4 4 4 (2421 out of 2421)
````

**Comment by jcupitt** [comment]
Huh how curious. Thanks for checking.

**Pull request merged**